src/sentry/utils/sdk.py

```diff
@@ -33,12 +34,33 @@ def is_current_event_safe():
     Tests the current stack for unsafe locations that would likely cause
     recursion if an attempt to send to Sentry was made.
     """
+
+    with configure_scope() as scope:
+        project_id = scope._tags.get("project")
+
+        if project_id and project_id == settings.SENTRY_PROJECT:
+            return False
+
     for _, filename, _, _, _, _ in inspect.stack():
         if filename.endswith(UNSAFE_FILES):
             return False
+
     return True
 
 
+def set_current_project(project_id):
+    """
+    Set the current project on the SDK scope for outgoing crash reports.
+
+    This is a dedicated function because it is also important for the recursion
+    breaker to work. You really should set the project in every task that is
+    relevant to event processing, or that task may crash ingesting
+    sentry-internal errors, causing infinite recursion.
+    """
+    with configure_scope() as scope:
+        scope.set_tag("project", project_id)
```

Comment:
i wonder if we can use this opportunity to also bind the org if the project as found in the cache.

Reply:
Would like to defer this discussion, @dcramer was also investing some time into consistent org tagging and I would like to consolidate that effort.

Reply:
Unrelated to the above: I definitely like the fact that we're passing in `projec_id`, and not the entire project. This will allow us to avoid loading the Project model in some cases, which will be quite a speedup when optimizing consumers.

Reply:
We could bind the org using the helpers I created, Id just worry it'd create more lookups (same concern as Jan)